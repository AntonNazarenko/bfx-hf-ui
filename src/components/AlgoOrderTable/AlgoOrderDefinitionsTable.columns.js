export default [{
  width: 300,
  label: 'Name',
  dataKey: 'name',
  cellRenderer: ({ rowData = {} }) => rowData.name,
},
{
  width: 300,
  label: 'Created',
  dataKey: 'mts',
  cellRenderer: ({ rowData = {} }) => {
    const timestamp = +rowData.mts / 1000
    return new Date(timestamp).toLocaleString()
  },
},
{
  width: 300,
  label: 'Type',
  dataKey: 'type',
  cellRenderer: ({ rowData = {} }) => rowData.orderType,
}, {
  width: 300,
  label: 'Amount',
  dataKey: 'amount',
  cellRenderer: ({ rowData = {} }) => rowData.amount,
},
{
  width: 300,
  label: 'Symbol',
  dataKey: 'symbol',
  cellRenderer: ({ rowData = {} }) => rowData.symbol,
}, {
  width: 300,
  label: 'Status',
  dataKey: 'status',
  cellRenderer: ({ rowData = {} }) => rowData.status,
}]
