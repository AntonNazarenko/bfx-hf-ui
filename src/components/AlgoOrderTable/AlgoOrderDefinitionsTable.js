import React from 'react'

import Panel from '../../ui/Panel'
import Table from '../../ui/Table'
import AlgoOrderDefinitionsTableColumns from './AlgoOrderDefinitionsTable.columns'
import { propTypes, defaultProps } from './AlgoOrderDefinitionsTable.props'

const ALGO_NAMES = {
  'bfx-accumulate_distribute': 'Accumulate/Distribute',
  'bfx-ping_pong': 'Ping/Pong',
  'bfx-iceberg': 'Iceberg',
  'bfx-twap': 'TWAP',
}

export default class AlgoOrderTable extends React.Component {
  static propTypes = propTypes

  static defaultProps = defaultProps

  shouldComponentUpdate() {
    return true
  }

  render() {
    const { orders, stopOrder, runOrder, algoOrders } = this.props
    const orderObjects = orders.reverse().map((ao) => {
      const isActive = !!ao[2]
      const isAOActive = algoOrders.filter( algo => algo.id === ao[1])[0].active
      let status = (
        <button
          type='button'
          className='hfui__stop-order-btn '
          onClick={() => runOrder(ao[0])}
        >
          Activate
        </button>
      )
      if (isActive) {
        status = (
          <button
            type='button'
            className='hfui__stop-order-btn '
            onClick={() => stopOrder(ao[0])}
          >
            Stop
          </button>
        )
      }
      if (!isAOActive) {
        status = (
          <div
            disabled={true}
            type='button'
            className='hfui__stop-order-btn-disabled'
          >
            {`${isActive ? 'Stop' : 'Inactivate' }`}
          </div>
        )
      }
      return {
        gid: ao[0],
        name: ALGO_NAMES[ao[1]],
        mts: ao[0],
        amount: ao[3].args.amount,
        orderType: ao[3].args.orderType,
        symbol: ao[3].args.symbol,
        price: ao[3].args.price,
        status,
      }
    })
    const noOrdersDialogue = (
      <div
        style={{
          textAlign: 'center',
        }}
      >
        <p style={{ marginTop: 70 }}>
          No algo orders have been created yet,
        </p>
        <p>Head to bitfinex.com and create one.</p>
      </div>
    )
    return (
      <Panel
        label={orders.length > 0 ? 'Orders' : ''}
        contentClassName='table__wrapper'
        style={{ height: '100%' }}
      >
        {
          (orders.length > 0) && (
            <Table
              data={orderObjects}
              columns={AlgoOrderDefinitionsTableColumns}
              maxWidth={850}
              defaultSortDirection='ASC'
            />
          )
        }
        {
          (orders.length <= 0) && (
            noOrdersDialogue
          )
        }
      </Panel>
    )
  }
}
