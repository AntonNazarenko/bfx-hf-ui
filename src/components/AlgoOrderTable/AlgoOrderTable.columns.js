import React from 'react'
import { Icon } from '@blueprintjs/core'
import Switch from 'react-switch'


export default [{
  width: 200,
  label: 'Name',
  dataKey: 'name',
  cellRenderer: ({ rowData = {} }) => rowData.name
}, {
  width: 200,
  label: 'Created',
  dataKey: 'mts',
  cellRenderer: ({ rowData = {} }) => (
    new Date(rowData.mts).toLocaleString()
  ),
}, {
  width: 80,
  label: 'Symbol',
  dataKey: 'symbol',
  cellRenderer: ({ rowData = {} }) => rowData.symbol
}, {
  width: 120,
  label: 'Order Count',
  dataKey: 'orderCount',
  cellRenderer: ({ rowData = {} }) => rowData.orderCount
}, {
  width: 120,
  label: 'Status',
  dataKey: 'status',
  cellRenderer: ({ rowData = {} }) => rowData.status 
}, {
  width: 120,
  label: 'Actions',
  dataKey: 'gid',
  cellRenderer: ({ rowData = {} }) => (
    rowData.status === 'ACTIVE'
      ? <Switch checked={true} height={14} width={28} onColor='#0F0' />
      : <Switch checked={false} height={14} width={28} offColor='#C1C2C3' />
  ),
}]
