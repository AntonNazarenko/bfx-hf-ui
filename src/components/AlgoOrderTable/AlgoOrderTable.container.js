import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'

import AlgoOrderActions from '../../redux/actions/algo-orders'
import EditorActions from '../../redux/actions/editor'

import AlgoOrderTableView from './AlgoOrderTable'

const mapStateToProps = (state = {}, ownProps = {}) => {
  const { orders = [] } = ownProps
  const { algoOrders = [] } = state.algoOrders
  const { editorOpened = false } = state.editor
  return {
    algoOrders,
    orders,
    editorOpened,
  }
}

const mapDispatchToProps = dispatch => ({
  changeStatus: (index) => {
    dispatch(AlgoOrderActions.changeStatus(index))
  },
  toggleEditor: (flag) => {
    dispatch(EditorActions.toggleEditor(flag))
  },
})

export default connect(mapStateToProps, mapDispatchToProps)(AlgoOrderTableView)
