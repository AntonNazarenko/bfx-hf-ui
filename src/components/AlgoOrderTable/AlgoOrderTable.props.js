import PropTypes from 'prop-types'

export const propTypes = {
  onSelect: PropTypes.func,
  orders: PropTypes.array,
  algoOrders: PropTypes.array,
}

export const defaultProps = {
  onSelect: () => {},
  orders: [],
  algoOrders: [],
}
