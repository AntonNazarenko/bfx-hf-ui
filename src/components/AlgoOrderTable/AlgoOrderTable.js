import React from 'react'

import Panel from '../../ui/Panel'
import Table from '../../ui/Table'
import AlgoOrderTableColumns from './AlgoOrderTable.columns'
import { propTypes, defaultProps } from './AlgoOrderTable.props'

const ALGO_NAMES = {
  'bfx-accumulate_distribute': 'Accumulate/Distribute',
  'bfx-ping_pong': 'Ping/Pong',
  'bfx-iceberg': 'Iceberg',
  'bfx-twap': 'TWAP',
}

export default class AlgoOrderTable extends React.Component {
  static propTypes = propTypes

  static defaultProps = defaultProps

  state = {
    editorOpened: false,
  }

  componentDidMount() {
    const { getTableData, algoOrders } = this.props
    // we can change this latter, if we need fetch data on each comonent mount
    const firstElementInRow = algoOrders[0][0]
    if (firstElementInRow === undefined) {
      getTableData()
    }
  }

  shouldComponentUpdate(nextProps) {
    const newAlgoOrders = nextProps.algoOrders
    const { algoOrders } = this.props

    if (newAlgoOrders.length !== algoOrders.length) {
      return true
    }

    const isSame = nextProps.algoOrders.every((row, indexRow) => row.every((value, indexValue) => value === algoOrders[indexRow][indexValue]))

    return !isSame
  }

  onRowClick({ index } = {}) {
    const {
      onSelect, algoOrders,
    } = this.props
    onSelect(algoOrders[index])
  }

  render() {
    const { algoOrders, orders } = this.props
    const orderObjects = algoOrders.map(ao => ({
      gid: ao[0],
      name: ALGO_NAMES[ao[1]],
      mts: ao[4],
      status: ao[2] ? 'ACTIVE' : 'STOPPED',
      orderCount: orders.filter(o => o[1] === +ao[0]).length,
      symbol: (orders.find(o => o[1] === +ao[0]) || {})[3] || '',
    }))

    return (
      <Panel label='Algo Orders' contentClassName='table__wrapper'>
        <Table
          data={orderObjects}
          columns={AlgoOrderTableColumns}
          onRowClick={e => this.onRowClick(e)}
          defaultSortBy='mts'
          defaultSortDirection='ASC'
        />
      </Panel>
    )
  }
}
