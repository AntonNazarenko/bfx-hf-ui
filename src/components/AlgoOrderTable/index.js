import AlgoOrderTable from './AlgoOrderTable.container'

export default AlgoOrderTable