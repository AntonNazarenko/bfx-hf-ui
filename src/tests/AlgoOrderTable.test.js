import React from 'react'
import renderer from 'react-test-renderer'

import AlgoOrderTable from '../components/AlgoOrderTable'

const algoOrders = [
  [42, 'bfx-ping_pong', true, null, +(new Date(Date.now() - (4 * 60 * 60 * 1000)))],
]

const orders = [
  // all we care about is gid & symbol for rendering ([_, gid, _, symbol])
  [null, 0, null, 'tBTCUSD'],
  [null, 0, null, 'tBTCUSD'],
  [null, 0, null, 'tBTCUSD'],
  [null, 0, null, 'tBTCUSD'],
  // note strat (1) is stopped
  [null, 2, null, 'tLEOUSD'],
  [null, 2, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
  [null, 42, null, 'tLEOUSD'],
]

test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <AlgoOrderTable
      algoOrders={algoOrders}
      orders={orders}
    />,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
