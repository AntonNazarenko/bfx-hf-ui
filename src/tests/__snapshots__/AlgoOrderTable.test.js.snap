// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<div
  className="panel__wrapper"
  style={undefined}
>
  <div
    className="panel__header"
    style={undefined}
  >
    <p>
      Algo Orders
    </p>
  </div>
  <div
    className="panel__content table__wrapper"
    style={undefined}
  >
    <div
      className={undefined}
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    >
      <div
        aria-colcount={6}
        aria-label={undefined}
        aria-labelledby={undefined}
        aria-rowcount={1}
        className="ReactVirtualized__Table"
        id={undefined}
        role="grid"
        style={Object {}}
      >
        <div
          className="ReactVirtualized__Table__headerRow"
          role="row"
          style={
            Object {
              "height": 24,
              "overflow": "hidden",
              "paddingRight": 0,
              "width": 0,
            }
          }
        >
          <div
            aria-label="Name"
            aria-sort="none"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Name"
            >
              Name
            </span>
          </div>
          <div
            aria-label="Created"
            aria-sort="ascending"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Created"
            >
              Created
            </span>
            <svg
              className="ReactVirtualized__Table__sortableHeaderIcon ReactVirtualized__Table__sortableHeaderIcon--ASC"
              height={18}
              viewBox="0 0 24 24"
              width={18}
            >
              <path
                d="M7 14l5-5 5 5z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
          <div
            aria-label="Symbol"
            aria-sort="none"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 80px",
                "flex": "0 1 80px",
                "msFlex": "0 1 80px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Symbol"
            >
              Symbol
            </span>
          </div>
          <div
            aria-label="Order Count"
            aria-sort="none"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 120px",
                "flex": "0 1 120px",
                "msFlex": "0 1 120px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Order Count"
            >
              Order Count
            </span>
          </div>
          <div
            aria-label="Status"
            aria-sort="none"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 120px",
                "flex": "0 1 120px",
                "msFlex": "0 1 120px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Status"
            >
              Status
            </span>
          </div>
          <div
            aria-label="Actions"
            aria-sort="none"
            className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
            id={undefined}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 120px",
                "flex": "0 1 120px",
                "msFlex": "0 1 120px",
              }
            }
            tabIndex={0}
          >
            <span
              className="ReactVirtualized__Table__headerTruncatedText"
              title="Actions"
            >
              Actions
            </span>
          </div>
        </div>
        <div
          aria-label="grid"
          aria-readonly={true}
          className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
          id={undefined}
          onScroll={[Function]}
          role="rowgroup"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "ltr",
              "height": -24,
              "outline": "none",
              "overflowX": "hidden",
              "overflowY": "auto",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
</div>
`;
